src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs

```diff
@@ -0,0 +1,97 @@
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.UseImplicitTyping
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseExplicitTyping), Shared]
+    internal class UseExplicitTypingCodeFixProvider : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds =>
+            ImmutableArray.Create(IDEDiagnosticIds.UseExplicitTypingDiagnosticId);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(span.Start);
+            if (!token.Span.IntersectsWith(span))
+            {
+                return;
+            }
+
+            var node = token.GetAncestors<SyntaxNode>().First(n => n.Span.Contains(span));
+            if (node == null ||
+                !(node.IsKind(SyntaxKind.PredefinedType) ||
+                  node.IsKind(SyntaxKind.ArrayType) ||
+                  node.IsKind(SyntaxKind.IdentifierName) ||
+                  node.IsKind(SyntaxKind.GenericName) ||
+                  node.IsKind(SyntaxKind.AliasQualifiedName)))
+            {
+                return;
+            }
+
+            var codeAction = new MyCodeAction(
+                CSharpFeaturesResources.UseExplicitTyping,
+                c => HandleDeclaration(document, root, node, context.CancellationToken));
+
+            context.RegisterCodeFix(codeAction, context.Diagnostics.First());
+        }
+
+        private async Task<Document> HandleDeclaration(Document document, SyntaxNode root, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declarationContext = node.Parent;
+
+            TypeSyntax typeSyntax;
+            if (declarationContext is VariableDeclarationSyntax)
+            {
+                typeSyntax = ((VariableDeclarationSyntax)declarationContext).Type;
+            }
+            else if (declarationContext is ForEachStatementSyntax)
+            {
+                typeSyntax = ((ForEachStatementSyntax)declarationContext).Type;
+            }
+            else
+            {
+                Debug.Assert(false, $"unhandled kind {declarationContext.Kind().ToString()}");
+                return null;
+            }
+
+            var typeSymbol = semanticModel.GetTypeInfo(typeSyntax).ConvertedType;
+
+            var typeName = SyntaxFactory.ParseTypeName(typeSymbol.ToDisplayString())
```

Comment:
Is there a place where anonymous types are filtered out by the diagnostic? I couldn't find it. Note: Anonymous types can appear in nefarious places, such as being passed through as generic type parameters. We have an extension method that checks for that [here] (http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/ITypeSymbolExtensions.cs,b2cea94feb884a75).